src/futures_future.rs

```diff
@@ -18,9 +18,34 @@ struct TaskWaker {
 	task: futures::task::Task,
 }
 
-impl ArcWake for TaskWaker {
-	fn wake(arc_self: &Arc<Self>) {
-		arc_self.task.notify();
+const TASK_WAKER_VTABLE : RawWakerVTable = RawWakerVTable::new(
+	clone_task_waker,
+	drop_task_waker,
+	wake_task_waker,
+);
+
+unsafe fn clone_task_waker(data: *const ()) -> RawWaker {
+	let data = &*(data as *const TaskWaker);
+	TaskWaker::into_raw_waker(data.task.clone())
+}
+
+unsafe fn drop_task_waker(data: *const ()) {
+	Box::from_raw(data as *mut TaskWaker);
+}
+
+unsafe fn wake_task_waker(data: *const ()) {
+	let data = &*(data as *const TaskWaker);
+	data.task.notify();
+}
+
+impl TaskWaker {
+	unsafe fn into_raw_waker(task: futures::task::Task) -> RawWaker {
+		let data = Box::leak(Box::new(task)) as *const _ as *const ();
```

Comment:
What's going on with the double `as ..`, is that necessary?

Reply:
Yeah, you can't cast directly to a pointer of a different type, but you can cast pointers to pointers of different types. So first it casts to a `*const futures::task::Task` (abbreviated as `*const _`, then to a `*const ()`.